crates/provider/src/layers/cache.rs

```diff
@@ -0,0 +1,539 @@
+    /// node.
+    fn get_accounts(&self) -> ProviderCall<T, NoParams, Vec<Address>> {
+        let req = RequestType::new("eth_accounts", NoParams::default());
+
+        let hash = req.params_hash().ok();
+
+        if let Some(hash) = hash {
+            if let Some(cached) = self.get(&hash).unwrap() {
+                let result = serde_json::from_str(&cached).map_err(TransportErrorKind::custom);
+                return ProviderCall::BoxedFuture(Box::pin(async move {
+                    let res = result?;
+                    Ok(res)
+                }));
+            }
+        }
+
+        let client = self
+            .inner
+            .weak_client()
+            .upgrade()
+            .ok_or_else(|| TransportErrorKind::custom_str("RPC client dropped"));
+        let cache = self.cache.clone();
+
+        ProviderCall::BoxedFuture(Box::pin(async move {
+            let client = client?;
+
+            let res = client.request_noparams(req.method()).await?;
+
+            let json_str = serde_json::to_string(&res).map_err(TransportErrorKind::custom)?;
+            let hash = req.params_hash()?;
+            let mut cache = cache.write();
+            let _ = cache.put(hash, json_str);
+            Ok(res)
+        }))
+    }
+
+    /// Gets the chain ID.
+    fn get_chain_id(&self) -> ProviderCall<T, NoParams, U64, u64> {
+        let req = RequestType::new("eth_chainId", NoParams::default());
+
+        let hash = req.params_hash().ok();
+
+        if let Some(hash) = hash {
+            if let Some(cached) = self.get(&hash).unwrap() {
+                let result = serde_json::from_str(&cached).map_err(TransportErrorKind::custom);
+                return ProviderCall::BoxedFuture(Box::pin(async move {
+                    let res = result?;
+                    Ok(res)
+                }));
+            }
+        }
+
+        let client = self
+            .inner
+            .weak_client()
+            .upgrade()
+            .ok_or_else(|| TransportErrorKind::custom_str("RPC client dropped"));
+        let cache = self.cache.clone();
+
+        ProviderCall::BoxedFuture(Box::pin(async move {
+            let client = client?;
+
+            let res = client
+                .request_noparams(req.method())
+                .map_resp(utils::convert_u64 as fn(U64) -> u64)
+                .await?;
+
+            let json_str = serde_json::to_string(&res).map_err(TransportErrorKind::custom)?;
+            let hash = req.params_hash()?;
+            let mut cache = cache.write();
+            let _ = cache.put(hash, json_str);
+            Ok(res)
+        }))
+    }
+
+    async fn get_block_by_number(
+        &self,
+        number: BlockNumberOrTag,
+        hydrate: bool,
+    ) -> TransportResult<Option<Block>> {
+        let hash = RequestType::new("eth_getBlockByNumber", (number, hydrate));
+
+        cache_get_or_fetch!(self, hash, self.inner.get_block_by_number(number, hydrate))
```

Comment:
this is problematic because of reorgs, we'd need to track chain height somehow, but out of scope for this pr imo